// curatrix-mcp-server/dist/index.js
import { access } from "node:fs/promises";
import path from "node:path";
import process from "node:process";
import { fileURLToPath } from "node:url";
import { z } from "zod";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { applyFixes, scanProject } from "curatrix-core";
const scanInputSchema = {
    path: z.string().optional(),
    format: z.enum(["json", "text"]).optional(),
};
const fixInputSchema = {
    path: z.string().optional(),
    issueIds: z.array(z.string()).optional(),
    autoConfirm: z.boolean().default(true),
};
const server = new McpServer({
    name: "curatrix-mcp-server",
    version: "0.1.0",
});
export async function runCuratrixScanTool({ path: targetPath, format, }) {
    const rootDir = path.resolve(targetPath ?? process.cwd());
    try {
        await access(rootDir);
    }
    catch {
        return {
            content: [
                {
                    type: "text",
                    text: JSON.stringify({
                        error: `Path does not exist: ${rootDir}`,
                    }, null, 2),
                },
            ],
            isError: true,
        };
    }
    try {
        const result = await scanProject({ rootDir });
        const payload = createScanPayload(result);
        return {
            content: [
                {
                    type: "text",
                    text: serializeScanResult(payload, format ?? "json"),
                },
            ],
        };
    }
    catch (error) {
        return {
            content: [
                {
                    type: "text",
                    text: JSON.stringify({
                        error: error instanceof Error ? error.message : String(error),
                    }, null, 2),
                },
            ],
            isError: true,
        };
    }
}
export async function runCuratrixFixTool({ path: targetPath, issueIds, autoConfirm, }) {
    const rootDir = path.resolve(targetPath ?? process.cwd());
    const validationError = await validateRootDir(rootDir);
    if (validationError) {
        return validationError;
    }
    try {
        const result = await scanProject({ rootDir });
        const issues = issueIds?.length
            ? result.issues.filter((issue) => issueIds.includes(issue.id) || issueIds.includes(issue.fingerprint))
            : result.issues.filter((issue) => issue.fixAvailability !== "none" || Boolean(issue.patch));
        const applyResult = await applyFixes({
            rootDir,
            issues,
            autoConfirm: autoConfirm ?? true,
        });
        return successResponse({
            rootDir,
            fixedCount: applyResult.applied.length,
            selectedIssues: issues.map((issue) => ({
                id: issue.id,
                ruleId: issue.ruleId,
                title: issue.title,
                source: issue.source,
                correctionContext: buildCorrectionContext(issue),
            })),
            changes: (applyResult.changes ?? []).map((change) => ({
                file: change.file,
                changeType: change.changeType,
                oldValue: change.oldValue,
                newValue: change.newValue,
                diff: change.diff,
                reasoning: toHumanReasoning(change.reasoning),
            })),
            skipped: applyResult.skipped,
        });
    }
    catch (error) {
        return errorResponse(error instanceof Error ? error.message : String(error));
    }
}
server.tool("curatrix_scan", "Scan the current project for security vulnerabilities using Curatrix.", scanInputSchema, runCuratrixScanTool);
server.tool("curatrix_fix", "Apply Curatrix fixes for selected issues in the current project.", fixInputSchema, runCuratrixFixTool);
if (isMainModule(import.meta.url)) {
    const transport = new StdioServerTransport();
    await server.connect(transport);
}
function serializeScanResult(result, format) {
    if (format === "text") {
        return [
            "Curatrix scan completed",
            "",
            JSON.stringify(result, null, 2),
        ].join("\n");
    }
    return JSON.stringify(result, null, 2);
}
function createScanPayload(result) {
    return {
        ...result,
        issues: result.issues.map((issue) => ({
            id: issue.id,
            ruleId: issue.ruleId,
            severity: issue.severity,
            description: issue.remediation ?? issue.title,
            category: issue.category,
            source: issue.source,
            file: issue.locations[0]?.file,
            line: issue.locations[0]?.line,
            correctionContext: buildCorrectionContext(issue),
        })),
    };
}
function buildCorrectionContext(issue) {
    return {
        type: correctionType(issue),
        action: correctionAction(issue),
        reasoning: correctionReasoning(issue),
        confidence: correctionConfidence(issue),
    };
}
function correctionType(issue) {
    if (issue.ruleId.startsWith("deps.")) {
        return "dependency-update";
    }
    if (issue.patch || /\.(js|ts|jsx|tsx|py|mjs|cjs)\b/.test(issue.locations[0]?.file ?? "")) {
        return "code-patch";
    }
    if (issue.fixAvailability === "apply" || issue.fixAvailability === "preview") {
        return "config-patch";
    }
    return "manual-review";
}
function correctionAction(issue) {
    const remediation = issue.remediation?.trim();
    if (remediation) {
        const npmInstallMatch = remediation.match(/npm install\s+[^\n]+/i);
        if (npmInstallMatch) {
            return npmInstallMatch[0];
        }
        return remediation;
    }
    switch (issue.ruleId) {
        case "infra.docker.missing-user":
            return "Add a non-root USER line to the Dockerfile.";
        case "secrets.env-not-ignored":
            return "Add .env to .gitignore.";
        case "ai.open-bind-address":
            return "Bind the service to 127.0.0.1 instead of 0.0.0.0.";
        default:
            return "Review this finding and apply the safest targeted fix.";
    }
}
function correctionReasoning(issue) {
    const packageEvidence = issue.evidence.find((entry) => entry.label === "package" || entry.label === "version")?.value;
    if (issue.ruleId === "deps.provider-vulnerability" || issue.ruleId === "deps.transitive-cve") {
        return `I recommend updating ${packageEvidence ?? "this dependency"} because the current version is tied to a reported vulnerability and the safer version reduces known risk.`;
    }
    if (issue.ruleId === "deps.suspicious-install-script") {
        return "I recommend reviewing this install script because it contains commands that commonly download, decode, or execute malicious payloads during install.";
    }
    if (issue.ruleId === "ai.unsafe-exec") {
        return "I recommend removing this dynamic execution path because it can let untrusted input run code directly.";
    }
    if (issue.remediation) {
        return toHumanReasoning(issue.remediation);
    }
    return `I recommend fixing this because ${issue.why.charAt(0).toLowerCase()}${issue.why.slice(1)}`;
}
function correctionConfidence(issue) {
    return Math.min(0.99, Math.max(0.5, Number(issue.confidence.toFixed(2))));
}
function toHumanReasoning(input) {
    const trimmed = input.trim();
    if (trimmed.length === 0) {
        return "I made this change because it reduces risk without making a broad or surprising modification.";
    }
    const sentence = trimmed.endsWith(".") ? trimmed : `${trimmed}.`;
    if (/^i\s+/i.test(sentence)) {
        return sentence;
    }
    return `I made this change because ${sentence.charAt(0).toLowerCase()}${sentence.slice(1)}`;
}
async function validateRootDir(rootDir) {
    try {
        await access(rootDir);
        return undefined;
    }
    catch {
        return errorResponse(`Path does not exist: ${rootDir}`);
    }
}
function successResponse(payload) {
    return {
        content: [
            {
                type: "text",
                text: JSON.stringify(payload, null, 2),
            },
        ],
    };
}
function errorResponse(message) {
    return {
        content: [
            {
                type: "text",
                text: JSON.stringify({ error: message }, null, 2),
            },
        ],
        isError: true,
    };
}
function isMainModule(moduleUrl) {
    return process.argv[1] ? path.resolve(process.argv[1]) === path.resolve(fileURLToPath(moduleUrl)) : false;
}
